app/src/main/java/api/NetworkModule.kt

```diff
@@ -0,0 +1,60 @@
+package api
+
+import com.example.practice.BuildConfig
+import com.google.gson.GsonBuilder
+import com.jakewharton.retrofit2.adapter.kotlin.coroutines.CoroutineCallAdapterFactory
+import kotlinx.coroutines.Deferred
+import okhttp3.OkHttpClient
+import okhttp3.logging.HttpLoggingInterceptor
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+import retrofit2.http.GET
+
+
+/**
+ * UserApi is service interface which deals with user list data
+ */
+interface UserApi {
+    /**
+     * Function returns list of user list items asynchronously using 'Deferred'
+     */
+    @GET("users")
+    fun fetchUsersList() : Deferred<List<UsersListItem?>?>
+}
+
+/**
+ * Singleton network module class provide retrofit instance and services
+ */
+object NetworkModule {
```

Comment:
Singletons are generally discouraged in programming. What makes the use of a singleton object here the right choice? 
Also discuss the difference between `object SomeObject {}` and `class SomeClass {}` and where one should be used over the other.

Reply:
@nets-dev
Q1] Singletons are generally discouraged in programming. What makes the use of a singleton object here the right choice?
**Ans** :- As user service API is a frequently used service, hence making a singleton object is the best choice to implement this scenario. If I make use of 'class' then for each and every request I have to create an instance and access user service API. To avoid memory overhead I have to make NetworkModule a singleton object.
Q2] Also discuss the difference between object SomeObject {} and class SomeClass {} and where one should be used over the other below.
**Ans** :- object - The 'object' is used when we have to implement a singleton. We get only a single reference throughout the app and can access member variable and functions with the same reference.
                 class - we can create multiple instances of the class. We need to create an instance to access member variables and functions.